py3status/modules/google_calendar.py

```diff
@@ -0,0 +1,438 @@
+        """
+        Builds the composite reponse to be output by the module by looping through all
+        events and formatting the necessary strings.
+
+        Returns: A composite containing the individual response for each event.
+        """
+        responses = []
+        index = 0
+
+        for event in self.events:
+            summary = event.get('summary')
+            location = event.get('location')
+            description = event.get('description')
+            url = event['htmlLink']
+
+            start_dt = gstr_to_datetime(event['start'].get('dateTime'))
+            end_dt = gstr_to_datetime(event['end'].get('dateTime'))
+
+            start_time_str = datetime_to_str(start_dt, self.format_time)
+            end_time_str = datetime_to_str(end_dt, self.format_time)
+
+            start_date_str = datetime_to_str(start_dt, self.format_date)
+            end_date_str = datetime_to_str(end_dt, self.format_date)
+
+            time_until = delta_time(start_dt)
+
+            self._check_warn_threshold(time_until, summary)
+
+            if time_until['days'] < 0:
+                time_until = delta_time(end_dt)
+                time_until_formatted = '{} {}'.format(self._format_timedelta(index, time_until),
+                                                      self.event_in_progress_suffix)
+            else:
+                time_until_formatted = self._format_timedelta(index, time_until)
+
+            self._check_button_open_url(index, url)
+
+            curr_event_format = self._check_button_toggle_format(index)
+
+            event_formatted = self.py3.safe_format(
+                curr_event_format,
+                {'summary': summary,
+                 'location': location,
+                 'description': description,
+                 'start_time': start_time_str,
+                 'end_time': end_time_str,
+                 'start_date': start_date_str,
+                 'end_date': end_date_str,
+                 'time_until': time_until_formatted})
+
+            responses.append({
+                'full_text': event_formatted,
+                'color': self.colors[index], 'index': index})
+
+            if index < self.num_events - 1:
+                responses.append({'full_text':
+                                  ' {} '.format(self.event_separator), 'index': 'sep'})
+
+            index += 1
+
+        return {'events': self.py3.composite_create(responses)}
+
+    def google_calendar(self):
+        """
+        The method that outputs the response.
+
+        First, we check credential authorization. If no authorization, we display an error message,
+        and try authorizing again in 5 seconds.
+
+        Otherwise, we fetch the events, build the response, and output the resulting composite.
+        """
+        if not self.is_authorized:
+            self.is_authorized = self._authorize_credentials()
+
+            error_msg = self.py3.safe_format(
+                self.format_error,
```

Comment:
I agree that `{error}` and `format_error` are unnecessary. The case I'm trying to cover here is when py3status loads at boot up before an internet connection is established. With `format_error` an empty string, as is the default now, the observed behaviour is that the module displays nothing, until successful authorization. It's similar to the behaviour of the weather_yahoo module under the same circumstances. 

So it's not something that happens often, just when the module is started without an internet connection available. 5 seconds is probably too often, so I could extend that to 30 or 60 seconds? I'll remove `{error}` and `format_error` if this approach sounds okay.

Reply:
I see. Maybe we should do a `boost run` here too... where we run with `cached_until` right away... because modules on `py3status` could be stuck waiting for this module to return a response... before we start seeing them all. This is a performance tweak... For me, it's worth having them in `gcal` and `coin_market` as both modules requires Internet.

Look at `coin_market` for `self.first_run` + `cached_until` for an example. You should notice some startup differences with this tweak.

It might be easier / cleaner to wait `60` for the refresh.... and users can refresh by clicking on the module. It's hard to tell without testing the module... as we often get the results from Internet not long after on the boot.

Reply:
It doesn't force other modules to wait on it because it does return a response if it can't get a connection...by default it just returns an empty string as the response, so the module appears as if it isn't in the bar at all until it gets some events and there is something to display.

I think this way is a bit friendlier for the user because they don't have to manually click to get the module working after an internet connection is established. I would be annoyed by that if I had to do it myself :P 

I'll wait and see until you get a chance to test it, and you let me know if the current behaviour isn't good. Maybe we can make it another configuration parameter where the user can decide if they want the module to keep trying if it can't establish a connection.